src/transformStateWithClones.js

```diff
@@ -7,7 +7,49 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  // const actionsCopy = actions;
+  const stateHistory = [];
+  const stateCopy = { ...state };
+
+  for (const key of actions) {
+    switch (key.type) {
+      case `addProperties`: {
+        Object.assign(stateCopy, key.extraData);
+
+        const newObj = structuredClone(stateCopy);
+
+        stateHistory.push(newObj);
+        break;
+      }
+
+      case `removeProperties`: {
+        for (const propertie of key.keysToRemove) {
+          delete stateCopy[propertie];
+        }
+
+        const newObj = structuredClone(stateCopy);
+
+        stateHistory.push(newObj);
+        break;
+      }
+
+      case `clear`: {
+        for (const propertie in stateCopy) {
+          delete stateCopy[propertie];
+        }
+
+        const newObj = structuredClone(stateCopy);
```

Comment:
Ensure that `structuredClone` is supported in the environment where this code will run. If not, consider using an alternative method for deep cloning.